keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+
+Each `ValidatingAdmissionPolicy` resource contains a `spec.match` to declare
+what resources it validates. This field is required.
+
+- `spec.match` constrains which resources this policy can be applied to. Policy
+  bindings each have match rules with further narrow this constraint, but cannot
+  expand it. This allows the CEL expressions to make safe assumptions. E.g. a
+  CEL expression that is constrained to CREATES and UPDATES of resources is
+  guaranteed the root `object` variable is never null, but a CEL expression that
+  might need to evaluate a DELETE must handle the root `object` variable being
+  null. See below "Match criteria" section for how match criteria is described.
+- `spec.match` guides CEL expression type checking, see the below "Type safety"
+  section for more details. 
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
+- `spec.match` - Match criteria for which resources the policy should validate
+- `spec.params` - Reference to the custom resource containing the params to use
+  when validating resources 
+- `spec.enforcement` - How validation failures should be enforced
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyBinding
+metadata:
+  name: "xyzlimit-scale.example.com"
+spec:
+  policy: xyzlimit-scale.example.com
+  params: xyzlimit-scale-settings.example.com
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny, Audit]
+```
+
+Each parameter CRD defines the custom resources that are referenced by the
+`spec.params` field of `PolicyBinding` resources. Example:
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: XyzLimit
+metadata:
+  name: "xyzlimit-settings.example.com"
+allowed: ["a", "b", "c"]
+banned: ["x", "y", "z"]
+xyz:
+  fuzzFactor: 0.8
+  reticulate: true
+```
+
+Note that for policies that require no parameterization, only the
+`PolicyBinding` is needed.
+
+See "Alternatives considered" section for rejected alternatives. This design was
+selected because:
+
+- Matching criteria is fully defined and validated in a builtin type.
+- Type checking is straight forward.
+- Policy parameterization is separated from the policy binding, allowing for
```

Comment:
I believe the above two points are also true if you have a (policy, binding) tuple where the binding contains the parameters inlined. Is there a benefit to being able to version the parameters type definition over time? I feel like perhaps there are other benefits which just haven't been enumerated.

Reply:
Combining binding and parameters gives you an object with two different owners for different parts of the schema, which mostly doesn't work (makes type safety very hard). All of the alternatives tried to do that in some way.

Reply:
The 'alternatives considered' has a bunch of binding + parameters inlined designs that didn't pan out. I've really warmed up to this (policy, binding, params) split over time. It makes some rollout workflows much easier (e.g. adding a new validation expression to a policy).

Reply:
@jpbetz I'm mostly looking for the benefits to be spelled out more clearly. If the schema ownership is a good point to separate concerns, then that's probably worth mentioning.

Reply:
SG, I've added this to the list as well as some other advantages that have been mentioned.